lib/events/session_writer.go

```diff
@@ -76,15 +77,12 @@ type AuditWriterConfig struct {
 	// Namespace is the session namespace.
 	Namespace string
 
-	// RecordOutput stores info on whether to record session output
-	RecordOutput bool
-
 	// Component is a component used for logging
 	Component string
 
 	// MakeEvents converts bytes written via the io.Writer interface
 	// into AuditEvents that are written to the stream.
-	// For backwards compatibility, AuditWriter will convert bytes to
+	// For backwards compatibility, SessionWriter will convert bytes to
 	// SessionPrint events when MakeEvents is not provided.
 	MakeEvents func([]byte) []apievents.AuditEvent
```

Comment:
Some of the bugs we had with things not getting recorded correctly were due to us mixing the `io.Writer` adapter in with a general purpose session recorder.

I wonder if this would be simpler if `SessionWriter` didn't implement the `io.Writer` interface and only exposed the generic `RecordEvent` method. Then we could have a separate adapter for cases where we need `io.Writer`.

I'm thinking something like:

```
// ByteStreamRecorder records an arbitrary stream of bytes by converting
// them to SessionPrint events.
type ByteStreamRecorder struct {
    recorder events.Recorder
}

func (b *ByteStreamRecorder) Write(b []byte) (n int, err error) {
    // package b up into one or more SessionPrint events
    b.recorder.RecordEvent(printEvent)
}
```

Reply:
Hmm, that's not a bad idea, however I think this could be made more simple by returning different interfaces from the `recorder` package. One function could return a session recorder interface without `io.Writer`, and another function could wrap it and return an interface that includes `io.Writer`.

That would avoid creating another arbitrary type and keep all session event logic in SessionWriter, while also ensuring callers only get access to functionality that they need.
```suggestion
	MakeEvents func([]byte) []apievents.AuditEvent
```

Reply:
After the changes to separate event setup and recording, doing this would necessitate adding 2 or more interfaces, a `SessionRecorder` without `io.Writer`, and `SessionSetterRecorder` without `io.Writer`. Not sure if this is worth making the event interface count even larger.